File: src/main/java/frc/robot/commands/Climber_Commands/AutoClimber.java
package frc.robot.commands.Climber_Commands;

import com.ctre.phoenix.motorcontrol.TalonFXControlMode;
import edu.wpi.first.wpilibj2.command.Command;
import frc.robot.Constants;
import frc.robot.subsystems.Climber;

import java.lang.Math;


public class AutoClimber extends Command {
    private double targetTraj = 0.0;
    private final Climber climber;
    private final boolean isGrenade;
    private final boolean isDown;

    public AutoClimber(Climber climber, boolean isGrenade, boolean isDown) {
        this.climber = climber;
        this.isGrenade = isGrenade;
        this.isDown = isDown;
        if(!isGrenade) {
            if (isDown) {
                targetTraj = Constants.CLIMBER_DOWN;
            } else {
                targetTraj = Constants.CLIMBER_UP;
            }
        } else {
            targetTraj = Constants.CLIMBER_GRENADE;
        }
        addRequirements(climber);
    }

    @Override
    public void initialize() {
        if (isGrenade) {
            Climber.isReleased = true;
            climber.resetEncoders();
            System.out.println("INFO: Grenade initialize");
        } else if (!Climber.isReleased) {
            System.out.println("WARNING: PULL GRENADE PIN!!!!!!!!!!");
        }
    }

    @Override
    public void execute() {
        climber.setAuto(TalonFXControlMode.MotionMagic, targetTraj);
    }

    @Override
    public boolean isFinished() {
        var traj_pose_error = targetTraj - climber.getEncoderPos();
        if (isGrenade) {
            return Math.abs(traj_pose_error) < 1000;
        } else {
            return false;
        }
    }

    @Override
    public void end(boolean interrupted) {
        System.out.println("INFO: Grenade end");
        climber.stop();
    }
}

File: src/main/java/frc/robot/commands/Climber_Commands/AutoClimber.kt
package frc.robot.commands.Climber_Commands

import com.ctre.phoenix.motorcontrol.TalonFXControlMode
import edu.wpi.first.wpilibj2.command.CommandBase
import frc.robot.Constants
import frc.robot.subsystems.Climber
import kotlin.math.abs

class AutoClimber(private val climber: Climber, private val isGrenade: Boolean, isDown: Boolean) : CommandBase() {
    private var targetTraj = 0.0

    init { addRequirements(climber)
        targetTraj = if(!isGrenade) {
            if (isDown) {
                Constants.CLIMBER_DOWN
            } else {
                Constants.CLIMBER_UP
            }
        } else {
            Constants.CLIMBER_GRENADE
        }
    }

    override fun initialize() {
        if (isGrenade) {
            Climber.isReleased = true
            climber.resetEncoders()
            println("INFO: Grenade initialize")
        } else if (!Climber.isReleased) {
            println("WARNING: PULL GRENADE PIN!!!!!!!!!!")
            return
        }
    }

    override fun execute() {
        climber.setAuto(TalonFXControlMode.MotionMagic, targetTraj)
    }

    override fun isFinished(): Boolean {
        val traj_pose_error = targetTraj - climber.getEncoderPos()
        return if (isGrenade) {
            abs(traj_pose_error) < 1000
        } else {
            false
        }
    }

    override fun end(interrupted: Boolean) {
        println("INFO: Grenade end")
        climber.stop()
    }

}
